grep/hemartin/grep.go

```diff
@@ -0,0 +1,94 @@
+package grep
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"regexp"
+)
+
+const printLinesFlag = "-n"
+const printNamesFlag = "-l"
+const caseInsensitiveFlag = "-i"
+const invertFlag = "-v"
+const matchLinesFlag = "-x"
+
+func Search(pattern string, flags, files []string) []string {
+	result := []string{}
+
+	printLines := sliceContains(flags, printLinesFlag)
+	caseInsensitive := sliceContains(flags, caseInsensitiveFlag)
+	printNames := sliceContains(flags, printNamesFlag)
+	matchLines := sliceContains(flags, matchLinesFlag)
+	invert := sliceContains(flags, invertFlag)
+
+	regexFlags := ""
+	if caseInsensitive {
+		regexFlags += "(?i)"
+	}
+
+	if matchLines {
+		pattern = fmt.Sprintf("^%s$", pattern)
+	}
+
+	regexPattern := fmt.Sprintf(`%s%s`, regexFlags, pattern)
+
+	compiledPattern := regexp.MustCompile(regexPattern)
+	filesContents := readFiles(files)
+	for _, fileName := range files {
+		lines := filesContents[fileName]
+		for lineIndex, line := range lines {
+			if string(line) == "" {
+				continue
+			}
+
+			match := compiledPattern.Match(line)
+			if (match && !invert) || (!match && invert) {
+
+				prefix := ""
+				if len(files) > 1 {
+					prefix += fmt.Sprintf("%s:", fileName)
+				}
+
+				if printLines {
+					prefix += fmt.Sprintf("%d:", lineIndex+1)
+				}
+
+				matchResult := fmt.Sprintf("%s%s", prefix, line)
+				if printNames {
+					if !sliceContains(result, fileName) {
+						result = append(result, fileName)
+					}
+				} else {
+					result = append(result, matchResult)
+				}
+			}
+		}
+	}
+
+	return result
+}
+
+func sliceContains(s []string, e string) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+func readFiles(files []string) map[string][][]byte {
```

Comment:
This method is not optimal as it gets information passed by value and it passes another structure of data again by value. Maybe a better way is to define a struct with the format of the return here value, have a method that fills the struct with data and pass pointers around.